rust/connlib/tunnel/src/tests/transition.rs

```diff
@@ -158,27 +162,44 @@ where
         })
 }
 
-pub(crate) fn dns_query<S>(
+/// Samples up to 5 DNS queries that will be sent concurrently into connlib.
+pub(crate) fn dns_queries(
     domain: impl Strategy<Value = DomainName>,
-    dns_server: impl Strategy<Value = S>,
-) -> impl Strategy<Value = Transition>
-where
-    S: Into<SocketAddr>,
-{
-    (
-        domain,
-        dns_server.prop_map_into(),
-        prop_oneof![Just(Rtype::A), Just(Rtype::AAAA)],
-        any::<u16>(),
-    )
-        .prop_map(
-            move |(domain, dns_server, r_type, query_id)| Transition::SendDnsQuery {
-                domain,
-                r_type,
-                query_id,
-                dns_server,
-            },
-        )
+    dns_server: impl Strategy<Value = SocketAddr>,
+) -> impl Strategy<Value = Vec<DnsQuery>> {
+    // Queries can be uniquely identified by the tuple of DNS server and query ID.
+    let unique_queries = collection::btree_set((dns_server, any::<u16>()), 1..5);
```

Comment:
Sampling from the full range of u16 makes it a bit unlikely that we actually hit the bug fixed in #6233. I temporarily set this to `0` locally and it also passes. It would be nice if we could somehow tell proptest that it should favor smaller values here or something to make a collision more likely.

I didn't want to arbitrarily use a smaller range though so I left it as `any::<u16>`.

Reply:
what about picking a random value and a small range around that value?

Reply:
How would you express that as a strategy that guarantees that you also only have unique pairs?

Can we merge this first and optimise later?

I think with the amount of CI runs across all platforms that we are doing, this isn't that unlikely. Picking the same DNS ID again happens with a probability of 4/65536, which is 1/16384. We run a 1000 testcases on each commit, not all are using DNS queries obviously but even if we say they are only involved in half of them, it is only 32 CI runs which probably happen in a day or so.

Reply:
> Can we merge this first and optimise later?

I'll make a note in the tracking issue so we don't forget.

Reply:
Yeah, not worth blocking on this :)

Reply:
Right now there's nothing really preventing a duplicate right? Only that it's unlikely.

Maybe-- If we reduce the sample space for ids to a 1000 around an arbitrary value seems like it's unlikely to have duplicates and we can filter for that case. And hopefully there won't be so many rejects and we make it considerably more likely to have an overlap.

Reply:
> Right now there's nothing really preventing a duplicate right? Only that it's unlikely.

Well, I am sampling a `BTreeSet` for the things that need to be unique. That guarantees that we don't sample something that we don't actually support.

> Maybe-- If we reduce the sample space for ids to a 1000 around an arbitrary value seems like it's unlikely to have duplicates and we can filter for that case. And hopefully there won't be so many rejects and we make it considerably more likely to have an overlap.

You mean like a two-step sampling? First sample a random u16 and then sample 5 IDs in a range of +- 100?